File: test/modules/linear_boltzmann_solvers/transport_adjoint/response_2d_1.lua
-- 2D Transport test with localized material source
-- SDM: PWLD
-- Test: QoI Value=1.38399e-05
--       Inner Product=1.38405e-05
num_procs = 4

-- Check num_procs
if check_num_procs == nil and number_of_processes ~= num_procs then
  log.Log(
    LOG_0ERROR,
    "Incorrect amount of processors. "
      .. "Expected "
      .. tostring(num_procs)
      .. ". Pass check_num_procs=false to override if possible."
  )
  os.exit(false)
end

-- Create mesh
N = 60
L = 5.0
ds = L / N

nodes = {}
for i = 0, N do
  nodes[i + 1] = i * ds
end

meshgen = mesh.OrthogonalMeshGenerator.Create({ node_sets = { nodes, nodes } })
grid = meshgen:Execute()

-- Set block IDs
grid:SetUniformBlockID(0)

vol1a = logvol.RPPLogicalVolume.Create({
  infx = true,
  ymin = 0.0,
  ymax = 0.8 * L,
  infz = true,
})

grid:SetBlockIDFromLogicalVolume(vol1a, 1, true)

vol0 = logvol.RPPLogicalVolume.Create({
  xmin = 2.5 - 0.166666,
  xmax = 2.5 + 0.166666,
  infy = true,
  infz = true,
})
grid:SetBlockIDFromLogicalVolume(vol0, 0, true)

vol2 = logvol.RPPLogicalVolume.Create({
  xmin = 2.5 - 0.166666,
  xmax = 2.5 + 0.166666,
  ymin = 0.0,
  ymax = 2 * 0.166666,
  infz = true,
})
grid:SetBlockIDFromLogicalVolume(vol2, 2, true)

vol1b = logvol.RPPLogicalVolume.Create({
  xmin = -1 + 2.5,
  xmax = 1 + 2.5,
  ymin = 0.9 * L,
  ymax = L,
  infz = true,
})
grid:SetBlockIDFromLogicalVolume(vol1b, 1, true)

-- Add cross sections to materials
xs_1g1 = xs.CreateSimpleOneGroup(0.01, 0.01)
xs_1g2 = xs.CreateSimpleOneGroup(0.1 * 20, 0.8)
xs_1g3 = xs.CreateSimpleOneGroup(0.3 * 20, 0.0)

-- Create sources
fwd_src = lbs.VolumetricSource.Create({ block_ids = { 2 }, group_strength = { 3.0 } })

-- Setup physics
pquad = aquad.CreateGLCProductQuadrature2DXY(12, 192)

lbs_block = {
  mesh = grid,
  num_groups = 1,
  groupsets = {
    {
      groups_from_to = { 0, 0 },
      angular_quadrature = pquad,
      inner_linear_method = "petsc_gmres",
      l_abs_tol = 1.0e-6,
      l_max_its = 500,
      gmres_restart_interval = 100,
    },
  },
  xs_map = {
    { block_ids = { 0 }, xs = xs_1g1 },
    { block_ids = { 1 }, xs = xs_1g2 },
    { block_ids = { 2 }, xs = xs_1g3 },
  },
  options = {
    scattering_order = 0,
    volumetric_sources = { fwd_src },
  },
}
phys = lbs.DiscreteOrdinatesSolver.Create(lbs_block)

-- Forward solve
ss_solver = lbs.SteadyStateSolver.Create({ lbs_solver = phys })

ss_solver:Initialize()
ss_solver:Execute()

-- Get field functions
ff_m0 = fieldfunc.GetHandleByName("phi_g000_m00")
ff_m1 = fieldfunc.GetHandleByName("phi_g000_m01")
ff_m2 = fieldfunc.GetHandleByName("phi_g000_m02")

-- Define QoI region
qoi_vol = logvol.RPPLogicalVolume.Create({
  xmin = 0.5,
  xmax = 0.8333,
  ymin = 4.16666,
  ymax = 4.33333,
  infz = true,
})

-- Compute QoI
ffi = fieldfunc.FieldFunctionInterpolationVolume.Create()
ffi:SetOperationType(OP_SUM)
ffi:SetLogicalVolume(qoi_vol)
ffi:AddFieldFunction(ff_m0)

ffi:Initialize(ffi)
ffi:Execute(ffi)
fwd_qoi = ffi:GetValue()

-- Create adjoint source
adj_src = lbs.VolumetricSource.Create({ logical_volume = qoi_vol, group_strength = { 1.0 } })

-- Switch to adjoint mode
adjoint_options = {
  adjoint = true,
  volumetric_sources = { adj_src },
}
phys:SetOptions(adjoint_options)

-- Adjoint solve, write results
ss_solver:Execute()
lbs.WriteFluxMoments(phys, "adjoint_2d_1")

-- Create response evaluator
buffers = { { name = "buff", file_prefixes = { flux_moments = "adjoint_2d_1" } } }
mat_sources = { { block_id = 2, strength = { 3.0 } } }
response_options = {
  lbs_solver = phys,
  options = {
    buffers = buffers,
    sources = { material = mat_sources },
  },
}
evaluator = lbs.ResponseEvaluator.Create(response_options)

-- Evaluate response
adj_qoi = evaluator:EvaluateResponse("buff")

-- Print results
log.Log(LOG_0, string.format("QoI Value=%.5e", fwd_qoi))
log.Log(LOG_0, string.format("Inner Product=%.5e", adj_qoi))

-- Cleanup
MPIBarrier()
if location_id == 0 then
  os.execute("rm adjoint_2d_1*")
end

File: test/python/modules/linear_boltzmann_solvers/transport_adjoint/response_2d_1.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
2D PWLD adjoint transport test with localized material source
Test: QoI Value=1.38399e-05
      Inner Product=1.38405e-05
"""

import os
import sys
import math

if "opensn_console" not in globals():
    from mpi4py import MPI
    size = MPI.COMM_WORLD.Get_size()
    rank = MPI.COMM_WORLD.Get_rank()
    barrier = MPI.COMM_WORLD.Barrier
    sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../../")))
    from pyopensn.mesh import OrthogonalMeshGenerator
    from pyopensn.xs import MultiGroupXS
    from pyopensn.source import VolumetricSource
    from pyopensn.aquad import GLCProductQuadrature2DXY
    from pyopensn.solver import DiscreteOrdinatesProblem, SteadyStateSourceSolver
    from pyopensn.response import ResponseEvaluator
    from pyopensn.fieldfunc import FieldFunctionInterpolationVolume
    from pyopensn.logvol import RPPLogicalVolume
else:
    barrier = MPIBarrier

if __name__ == "__main__":

    # Check number of processors
    num_procs = 4
    if size != num_procs:
        sys.exit(f"Incorrect number of processors. Expected {num_procs} processors but got {size}.")

    # Create mesh
    N = 60
    L = 5.0
    ds = L / N
    nodes = []
    for i in range(N + 1):
        nodes.append(i * ds)
    meshgen = OrthogonalMeshGenerator(node_sets=[nodes, nodes])
    grid = meshgen.Execute()
    grid.SetUniformBlockID(0)

    # Define a logical volume that spans in x (infinite) and in y from 0 to 0.8*L.
    vol1a = RPPLogicalVolume(infx=True, ymin=0.0, ymax=0.8 * L, infz=True)
    grid.SetBlockIDFromLogicalVolume(vol1a, 1, True)

    # Define a volume for block 0: localized in x around 2.5.
    vol0 = RPPLogicalVolume(xmin=2.5 - 0.166666, xmax=2.5 + 0.166666, infy=True, infz=True)
    grid.SetBlockIDFromLogicalVolume(vol0, 0, True)

    # Define a volume for block 2: localized in x and in y from 0 to 2*0.166666.
    vol2 = RPPLogicalVolume(
        xmin=2.5 - 0.166666,
        xmax=2.5 + 0.166666,
        ymin=0.0,
        ymax=2 * 0.166666,
        infz=True,
    )
    grid.SetBlockIDFromLogicalVolume(vol2, 2, True)

    # Define a second volume for block 1: localized in x around 2.5 and in y from 0.9*L to L.
    vol1b = RPPLogicalVolume(
        xmin=2.5 - 1,  # equivalent to (-1 + 2.5)
        xmax=2.5 + 1,  # equivalent to (1 + 2.5)
        ymin=0.9 * L,
        ymax=L,
        infz=True,
    )
    grid.SetBlockIDFromLogicalVolume(vol1b, 1, True)

    # Add cross sections to materials
    xs_1g1 = MultiGroupXS()
    xs_1g1.CreateSimpleOneGroup(0.01, 0.01)

    xs_1g2 = MultiGroupXS()
    xs_1g2.CreateSimpleOneGroup(0.1 * 20, 0.8)

    xs_1g3 = MultiGroupXS()
    xs_1g3.CreateSimpleOneGroup(0.3 * 20, 0.0)

    # Create sources
    # Forward source: only active in block 2 with a group strength of 3.0.
    fwd_src = VolumetricSource(block_ids=[2], group_strength=[3.0])

    # Create a 2D angular quadrature with 12 polar and 192 azimuthal angles.
    pquad = GLCProductQuadrature2DXY(n_polar=12, n_azimuthal=192, scattering_order=0)

    # Setup solver
    phys = DiscreteOrdinatesProblem(
        mesh=grid,
        num_groups=1,
        groupsets=[
            {
                "groups_from_to": (0, 0),
                "angular_quadrature": pquad,
                "inner_linear_method": "petsc_gmres",
                "l_abs_tol": 1.0e-6,
                "l_max_its": 500,
                "gmres_restart_interval": 100,
            },
        ],
        xs_map=[
            {"block_ids": [0], "xs": xs_1g1},
            {"block_ids": [1], "xs": xs_1g2},
            {"block_ids": [2], "xs": xs_1g3},
        ],
        scattering_order=0,
        volumetric_sources=[fwd_src],
    )

    # Forward solve
    ss_solver = SteadyStateSourceSolver(problem=phys)
    ss_solver.Initialize()
    ss_solver.Execute()

    # Field functions
    fflist = phys.GetScalarFieldFunctionList(only_scalar_flux=False)

    # Define QoI region and compute forward QoI
    qoi_vol = RPPLogicalVolume(xmin=0.5, xmax=0.8333, ymin=4.16666, ymax=4.33333, infz=True)
    ffi = FieldFunctionInterpolationVolume()
    ffi.SetOperationType("sum")
    ffi.SetLogicalVolume(qoi_vol)
    ffi.AddFieldFunction(fflist[0][0])
    ffi.Initialize()
    ffi.Execute()
    fwd_qoi = ffi.GetValue()

    # Create adjoint source and switch to adjoint mode
    adj_src = VolumetricSource(logical_volume=qoi_vol, group_strength=[1.0])
    adjoint_options = {
        "adjoint": True,
        "volumetric_sources": [adj_src],
    }
    phys.SetOptions(
        adjoint=True
    )
    phys.SetVolumetricSources(volumetric_sources=[adj_src])

    # Adjoint solve and write flux moments
    ss_solver.Execute()
    phys.WriteFluxMoments("adjoint_2d_1")

    # Create response evaluator and evaluate response
    evaluator = ResponseEvaluator(problem=phys)
    evaluator.SetOptions(
        buffers=[{'name': 'buff', 'file_prefixes': {'flux_moments': 'adjoint_2d_1'}}],
        sources={'material': [{'block_id': 2, 'strength': [3.0]}]}
    )
    adj_qoi = evaluator.EvaluateResponse("buff")

    # Print results
    if rank == 0:
        print(f"QoI Value={fwd_qoi:.5e}")
        print(f"Inner Product={adj_qoi:.5e}")

    # Cleanup
    barrier()
    if rank == 0:
        os.system("rm adjoint_2d_1*")
